Add FATES land use change module to pass LUH2 data to FATES

glemieux opened the pull request:
This pull request adds a new module, the structure of which is adopted from the dyn_subgrid code, 
to enable e3sm to ingest minimally processed luh2 data to be passed to FATES.  
A new namelist variable is introduced `use_fates_luh`, to engage the module functionality, 
which is off by default.  The luh2 dataset to be ingested is a minimally processed concatenation 
of the luh2 state, transition and management datasets.

These changes also include a minor bug fix discovered in the process of developing the code, 
in which if the number of FATES patches being defined by the user in the fates parameter file is greater than the `max_patch_per_col` elm variable, would result in a BalanceCheck error.  

To be coordinated with https://github.com/NGEET/fates/pull/1040

[nonBFB] for FATES

bishtgautam commented on components/elm/src/dyn_subgrid/dynFATESLandUseChangeMod.F90:
```diff
@@ -0,0 +1,194 @@
+    landuse_states = 0._r8
+    landuse_transitions = 0._r8
+
+    if (use_fates_luh) then
+
+    ! Generate the dyn_file_type object
+    ! TO DO: check whether to initialize with start or end
+    dynFatesLandUse_file = dyn_file_type(landuse_filename, YEAR_POSITION_START_OF_TIMESTEP)
+    ! dynFatesLandUse_file = dyn_file_type(landuse_filename, YEAR_POSITION_END_OF_TIMESTEP)
+
+    ! Get initial land use data
+    num_points = (bounds%endg - bounds%begg + 1)
+    landuse_shape(1) = num_points ! Does this need an explicit array shape to be passed to the constructor?
+    do varnum = 1, num_landuse_transition_vars
+       landuse_transition_vars(varnum) = dyn_var_time_uninterp_type( &
+            dyn_file=dynFatesLandUse_file, varname=landuse_transition_varnames(varnum), &
+            dim1name=grlnd, conversion_factor=1.0_r8, &
+            do_check_sums_equal_1=.false., data_shape=landuse_shape)
+    end do
+    do varnum = 1, num_landuse_state_vars
+       landuse_state_vars(varnum) = dyn_var_time_uninterp_type( &
+            dyn_file=dynFatesLandUse_file, varname=landuse_state_varnames(varnum), &
+            dim1name=grlnd, conversion_factor=1.0_r8, &
+            do_check_sums_equal_1=.false., data_shape=landuse_shape)
+    end do
+    end if
```
Please indent this code block.

glemieux replied:
completed

bishtgautam commented on components/elm/src/main/elmfates_interfaceMod.F90:
```diff
@@ -995,6 +1019,15 @@ subroutine dynamics_driv(this, bounds_clump, top_as_inst,          &
          this%fates(nc)%bc_in(s)%site_area=col_pp%wtgcell(c)*grc_pp%area(g)*m2_per_km2
 
+         if (use_fates_luh) then
+            ! if (do_landuse_update) then
```
Please delete the commented out code and indent the code.

glemieux replied:
done

bishtgautam commented:
@glemieux Can we add a new test for this PR?

glemieux commented:
Status update: there is a forthcoming update to the fates luh2 data pipeline https://github.com/NGEET/fates/pull/1032 that will result in an update to the default fates landuse data set.  As such I've added "WIP" to this PR.

rljacob commented:
status:  probably another month waiting on FATES refactor.

rljacob commented:
Need #5849 and also still WIP.

rljacob commented:
#5849 was merged.  Status now?

rljacob commented:
@glemieux is this ready?

glemieux commented:
> @glemieux is this ready?

This is still WIP and waiting on https://github.com/NGEET/fates/pull/1040  updates.  I'm actively working on this.

rljacob commented:
update: still waiting on fates PR.  Another couple of weeks.

rljacob commented:
update:  still WIP.  2 others will need to be merged before this.

rljacob commented:
@glemieux can you list in a comment the PRs this is waiting on?

glemieux commented:
@rljacob it needs to be staged after #6018 and #6027, both which need review and are undergoing some testing.  Ultimately, it's also waiting on https://github.com/NGEET/fates/pull/1040, but this should be integrated within the next week.  

This PR needs to have some updates brought in, but is close to not being "WIP".

rgknox commented on components/elm/src/biogeophys/SurfaceAlbedoMod.F90:
```diff
@@ -754,13 +754,16 @@ subroutine SurfaceAlbedo(bounds,     &
-          p = filter_vegsol(fp)
-          rho(p,ib) = max( rhol(veg_pp%itype(p),ib)*wl(p) + rhos(veg_pp%itype(p),ib)*ws(p), mpe )
-          tau(p,ib) = max( taul(veg_pp%itype(p),ib)*wl(p) + taus(veg_pp%itype(p),ib)*ws(p), mpe )
+    ! rho and tau are not needed if fates is on 
```
great, glad this fix is in

rgknox commented on components/elm/src/dyn_subgrid/dynFATESLandUseChangeMod.F90:
```diff
@@ -0,0 +1,194 @@
+       ! Generate the dyn_file_type object
+       ! TO DO: check whether to initialize with start or end
+       dynFatesLandUse_file = dyn_file_type(landuse_filename, YEAR_POSITION_START_OF_TIMESTEP)
+       ! dynFatesLandUse_file = dyn_file_type(landuse_filename, YEAR_POSITION_END_OF_TIMESTEP)
```
repeated text, ok to remove?

glemieux replied:
Totally; Removed.

rgknox reviewed (commented):
@glemieux and I are discussing some cosmetic things and checking on a setting in the namelist build perl code, but otherwise I think this PR looks good.  @glemieux plans to add a test for this as well

rgknox commented:
FYI: The next FATES API (33) is fairly simple, but I will hold on making that until this branch is re-based, since I have to build it on top of this one.

glemieux commented:
- [ ] add landuse testmod to `fates` test suite
- [ ] remove `use_cn` return check from landusemod
- [ ] add `fluh_timeseries` to `setup_cmdl_fates_mode`

glemieux commented:
@peterdschwartz @rgknox I've taken care of the suggested changes and  rebased this onto `master`.  I'm going to start regression testing on this today and I've removed the '[WIP]'.

glemieux commented:
@peterdschwartz regression testing on perlmutter with the `e3sm_land_developer` suite is complete and all expected non-fates tests are B4B.  The `qian` smoke test from #6192 is failing, but that's because I hadn't rebased since last week.  I've got a local rebased branch with #6219 included to make sure its not failing for a different reason.  All fates test pass with expected DIFFs in non-landuse run modes since the tag update also includes a couple fates scientific bug fixes since `sci.1.68.2_api.31.0.0`.

glemieux commented:
Just a quick update to note that rebasing against the latest `master` resulted in `SMS.r05_r05.I1850ELMCN.pm-cpu_....elm-qian_1948` passing.

peterdschwartz commented:
Tested on chrysalis and everything came back as expected.  This can go in once it's got one approval

bishtgautam commented on components/elm/bld/namelist_files/namelist_defaults.xml:
```diff
@@ -565,6 +565,11 @@ lnd/clm2/surfdata_map/surfdata_conusx4v1_simyr2000_c160503.nc</fsurdat>
 
+<!-- Land Use Harmonization unified data sets for dynamic FATES land use change -->
+<fluh_timeseries hgrid="4x5" sim_year_range="1850-2015" use_fates=".true."
+ >lnd/clm2/surfdata_map/fates-sci.1.68.3_api.31.0.0_tools.1.0.1/LUH2_states_transitions_management.timeseries_4x5_hist_simyr1850-2015_c231101.nc</fluh_timeseries>
```
1. This file is not on the E3SM input data server.
2. FATES's SCI/API version number will continue to get updated. This would mean we would have many `fates-sci*` directories. How we have a single dir (e.g. `lnd/clm2/surfdata_map/fates`) and the metadata of the LUH2 file can include the information regarding the SCI/API version. In the F90 code, you can also include a sanity check regarding the SCI/API version.

glemieux replied:
1. @bishtgautam this file is fairly large.  I submitted a request about 2 weeks ago to get access to the ANL servers using the procedure laid out on confluence so that I could load this and future updates directly.  I haven't seen movement on that request.  In the meantime, the data is on `perlmutter`; is this something you could upload for me via globus?
2.  That should work.  The file structure here was the recommended by ctsm folks, although I can't remember the rationale currently.  I'll definitely add the fates sci/api version to the data file meta data.  That said, the meta data check isn't really needed for usage here.

bishtgautam commented on components/elm/bld/namelist_files/namelist_defaults.xml:
```diff
@@ -127,7 +127,7 @@ attributes from the config_cache.xml file (with keys converted to upper-case).
 <!-- FATES default parameter file                                       -->
 <!-- ================================================================== -->
-<fates_paramfile >lnd/clm2/paramdata/fates_params_api25.5.0_12pft_c230710.nc</fates_paramfile>
+<fates_paramfile >lnd/clm2/paramdata/fates_params_api.32.0.0_12pft_c231215.nc</fates_paramfile>
```
This file is not available on https://web.lcrc.anl.gov/public/e3sm/inputdata/lnd/clm2/paramdata/

glemieux replied:
Thanks @bishtgautam I'll email you the file.

peterdschwartz replied:
The file is on the other server though.  https://svn-ccsm-inputdata.cgd.ucar.edu/trunk/inputdata/lnd/clm2/paramdata/

bishtgautam replied:
Oh, FATES also contributes code to CTSM, so it is on NCAR's server. In that case, we don't need to put on the E3SM server.

glemieux replied:
Oh nice, I didn't realize that!  Thanks @peterdschwartz

bishtgautam commented on components/elm/src/dyn_subgrid/dynFATESLandUseChangeMod.F90:
```diff
@@ -0,0 +1,187 @@
+  type(dyn_file_type), target   :: dynFatesLandUse_file
+
+  ! Land use name arrays
+  character(len=5), public, parameter  :: landuse_state_varnames(num_landuse_state_vars) = &
+                    [character(len=5)  :: 'primf','primn','secdf','secdn','pastr','range', &
+                                          'urban','c3ann','c4ann','c3per','c4per','c3nfx']
```
How about adding the long name for these in the comment?

```
  character(len=5), public, parameter  :: landuse_state_varnames(num_landuse_state_vars) = &
                    [character(len=5)  :: 'primf', & ! Primary forest 
                                          'primn', & ! Primary something???
```

glemieux replied:
That's a good idea.  I could also point developers to the [data readme](https://luh.umd.edu/LUH2/LUH2_v2h_README.pdf) location online.  How do we feel about having urls in the code comments?

bishtgautam replied:
I believe the URLs are discouraged because they would inevitably get broken in the future.

glemieux replied:
Updated via f09f5e3b8bf54ce5eb13e7597946cfbcf4a7f39e

bishtgautam reviewed (changes requested):
I have a few minor comments. However, a major comment is: Shouldn't there be a new compset (say I20TRELMFATES) in which the transient LU data is used by FATES?

glemieux commented:
> However, a major comment is: Shouldn't there be a new compset (say I20TRELMFATES) in which the transient LU data is used by FATES?

Eventually we should have a compset for this, but with this PR I don't think its necessary as we're going to be bringing in a fates landuse v2 change in the next month or so.  I think with the v2 update we will want to facilitate easier use of the option through the compset.  

On a related note and by comparison, I think its definitely worth creating a fates SP mode compset with a future PR since that run mode is more mature.

peterdschwartz commented:
Still waiting for the repo to be open.
